File: scripts/check_tokenizer_js.mjs
/**
 * Assert the browser tokenizer matches the Python one.
 *
 * Fixtures are written by scripts/check_tokenizer_js.py: real COCO captions
 * plus the ids CaptionTokenizer produced for them. Any divergence in the BPE
 * merges, the byte mapping or the pre-tokenizer regex shows up as a mismatch.
 *
 * Usage: node scripts/check_tokenizer_js.mjs <fixtures.json>
 */
import { readFileSync } from "node:fs";
import { ByteLevelBPETokenizer } from "../demo/static/tokenizer.js";

const fixtures = JSON.parse(readFileSync(process.argv[2], "utf8"));
const tok = new ByteLevelBPETokenizer(JSON.parse(readFileSync(fixtures.spec_path, "utf8")));

let checked = 0;
const failures = [];
for (const { text, ids, mask } of fixtures.cases) {
  const got = tok.encode(text, fixtures.max_len);
  if (JSON.stringify(got.ids) !== JSON.stringify(ids) ||
      JSON.stringify(got.mask) !== JSON.stringify(mask)) {
    if (failures.length < 5) failures.push({ text, expected: ids.slice(0, 12), got: got.ids.slice(0, 12) });
  }
  checked++;
}

if (failures.length) {
  console.error(`MISMATCH on ${failures.length}+ of ${checked} cases`);
  for (const f of failures) console.error(JSON.stringify(f));
  process.exit(1);
}
console.log(`tokenizer parity ok — ${checked} cases, vocab ${tok.vocabSize}`);

File: scripts/check_tokenizer_js.py
#!/usr/bin/env python
"""Write tokenizer fixtures, then run the JS parity check against them."""
from __future__ import annotations

import json
import subprocess
import sys
import tempfile
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1] / "src"))
from microclip.config import load_config  # noqa: E402
from microclip.data.tokenizer import CaptionTokenizer  # noqa: E402

REPO = Path(__file__).resolve().parents[1]


def main() -> None:
    cfg = load_config(REPO / "configs" / "base.yml")
    spec_path = REPO / cfg["tokenizer"]["path"]
    max_len = cfg["data"]["max_text_len"]
    tokenizer = CaptionTokenizer(str(spec_path))

    with open(REPO / cfg["data"]["root"] / cfg["data"]["val_ann"]) as f:
        captions = [a["caption"] for a in json.load(f)["annotations"]]

    # Real captions, plus edge cases the demo will actually see from users.
    texts = captions[:3000] + [
        "", " ", "a", "A MAN.", "  leading spaces", "trailing  ",
        "émoji café naïve", "日本語のテキスト", "🐶 dog emoji 🏄",
        "hyphen-ated, punctuation!? 42 numbers 3.14",
        "a " * 80,  # longer than max_len: must truncate identically
    ]
    cases = []
    for text in texts:
        ids, mask = tokenizer.encode(text, max_len)
        cases.append({"text": text, "ids": ids, "mask": mask})

    with tempfile.NamedTemporaryFile("w", suffix=".json", delete=False) as f:
        json.dump({"spec_path": str(spec_path), "max_len": max_len, "cases": cases}, f)
        fixtures = f.name

    result = subprocess.run(["node", str(REPO / "scripts" / "check_tokenizer_js.mjs"), fixtures])
    Path(fixtures).unlink()
    sys.exit(result.returncode)


if __name__ == "__main__":
    main()
